### DISCUSSION TITLE
[gatsby-source-contentful] Why do content types use name instead of ID?

### DISCUSSION THREAD

**Discussion by njbmartin** [gatsbyjs/gatsby#34400] [Ideas / Feature Requests]
@axe312ger I've just stumbled on an issue when renaming the "display name" for a content type, which results in queries suddenly failing because the type this plugin creates uses the name field (eg. `Content Highlight Section`), and not the ID (eg. `contentHighlightSection`).

I have prefixed some content types with `[deprecated]` to show content editors which types not to use any more without their content suddenly disappearing... unfortunately that's exactly what has happened here. If I wasn't pinning the schema with `gatsby-plugin-schema-snapshot`, this may not have surfaced as an issue, but I definitely believe the type name should be inferred from the ID.

The name field can be changed at any time, whereas the ID field cannot, so I think the below doesn't make sense, at least for the type definition.

https://github.com/gatsbyjs/gatsby/blob/6ffe0afd15a4f4b327a108399f14183f337eab97/packages/gatsby-source-contentful/src/create-schema-customization.js#L117-L138

I can't use the config option `useNameForId: false`, as that changes the content nodes, so all the queries then break. Also, `useNameForId` when it's a `name` field?

**Comment by njbmartin** [comment]
Correction, it breaks queries regardless if you use fragments ;)

**Reply by axe312ger** [reply]
Sorry for the inconvenience, this pattern was introduced as IDs can not be changed in Contentful but these IDs might conflict with the GatsbyJS naming scheme.

What you can do:

1. Keep using IDs which should not change queries at all
2. Do not rename the content type names when using `useNameForId`
3. Or do rename and rewrite all queries :/

Alternatively: To get rid of this madness, I am working hard on #31385. As this is a huge change to the plugin, development on that new major will continue but take some more time.
